### Linear Regression with Coefficient and RMSE

## Merge feature and test datasets

In [1]:
### Importing the required packages and libraries
# we will need numpy and pandas later
import numpy as np
import pandas as pd


from sklearn.linear_model import LinearRegression
#import DecisionTreeClassifier class
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
# Randomly splitting the original dataset into training set and testing set
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

In [2]:
# reading a CSV file directly from Web, and store it in a pandas DataFrame:
# "read_csv" is a pandas function to read csv files from web or local device:
# walmart_features_df =  pd.read_csv('features.csv')

# walmart_sampleSubmission_df =  pd.read_csv('sampleSubmission.csv')

# walmart_stores_df =  pd.read_csv('stores.csv')

# walmart_test_df =  pd.read_csv('test.csv')

# walmart_train_df =  pd.read_csv('train.csv')

In [3]:
df = pd.read_csv('../data/merged-train-data.csv')
df[0::100000]

Store  Dept  Weekly_Sales  IsHoliday  Temperature  Fuel_Price  \
0           1     1      24924.50      False        42.31       2.572   
100000     11    91      40968.40      False        84.94       2.577   
200000     21    52       2314.78      False        68.61       2.720   
300000     31    67       5729.66      False        69.79       3.274   
400000     43    91      30713.10      False        54.96       2.771   

               CPI  Unemployment Type    Size  Markdowns  Year  Month  Day  
0       211.096358         8.106    A  151315        0.0  2010      2    5  
100000  214.848169         7.346    A  207499        0.0  2010      9    3  
200000  211.471329         8.163    B  140167        0.0  2010     10   15  
300000  217.004826         7.441    A  203750        0.0  2011     10   14  
400000  203.819129        10.210    C   41062        0.0  2010     11   19

### Split the merged data set into X_train and y_train, use test dataset for X_test

In [4]:
#split with method

y =df['Weekly_Sales']
X = df.drop(columns=['Weekly_Sales','Type'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, 
                                                    random_state=2)

In [5]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(X_train.head())

(295099, 12)
(295099,)
(126471, 12)
        Store  Dept  IsHoliday  Temperature  Fuel_Price         CPI  \
154115     16    98      False        31.73       3.479  195.718471   
238177     25    81      False        57.17       2.958  203.877024   
407348     44    91      False        23.35       2.974  127.719581   
6325        1    28      False        71.74       3.274  217.355273   
132096     14    22      False        45.01       3.582  184.613419   

        Unemployment    Size  Markdowns  Year  Month  Day  
154115         6.232   57197    7502.89  2011     11   18  
238177         7.856  128107       0.00  2010      5   21  
407348         7.224   39910       0.00  2011      2    4  
6325           7.866  151315       0.00  2011     10   14  
132096         8.549  200898       0.00  2011      3   11  


In [6]:
#Linear Regression

linear_reg = LinearRegression()
linear_reg.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [7]:
print("Intercept: ", linear_reg.intercept_)

print("Coefficient: " , linear_reg.coef_)

Intercept:  1366427.6184769098
Coefficient:  [-8.73094172e+01  1.10841005e+02  7.05852062e+02  1.98428062e+01
  2.92467634e+02 -2.08583897e+01 -1.83231124e+02  8.58850479e-02
  3.79210892e-02 -6.77895285e+02  1.30091858e+02 -1.54005316e+01]


### predictive model

$$y = 1366427.618 - 87.309 \times Store + 101.841 \times Department + 705.852 \times Is Holiday+ 19.842 \times Temperature + 292.467 \times FuelPrice - 20.858 \times CPI - 183.231 \times Unemployment + 0.0859 \times Size + 0.0379 \times Markdowns - 677.895 \times Year + 130.009 \times Month - 15.400 \times Day $$

###  Most important feature is 'IsHoliday'
### meaning that whether that specific day is a holiday or not has the greatest impact on weekly sales
### Second most important is Fuel Price follwed by Unemployment
### least important is markdowns but a lot of the data was 0

### Testing and prediction

In [8]:
y_pred = linear_reg.predict(X_test)
print(y_pred)

[10912.66259985 22031.38094417 18226.68931311 ...  4949.20422364
 17206.18091831 18621.10256506]


### RMSE

In [9]:
from sklearn import metrics
mse = metrics.mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)

In [54]:
print(rmse)

21830.539350501567
